### PR TITLE
feat: Refer to other packages from same `pkgsDirectory`

### LINKED ISSUE #1: Refer to other packages from same `pkgsDirectory`
Packages that try to use other packages in the same `pkgsDirectory` fail with the error:

```
$ nix build .#b
error:
       … while evaluating the attribute 'optionalValue.value'
         at /nix/store/x3d704r2qzq7h0m7i04liyzrvnng1pi2-source/modules.nix:868:5:
          867|
          868|     optionalValue =
             |     ^
          869|       if isDefined then { value = mergedValue; }

       … while evaluating a branch condition
         at /nix/store/x3d704r2qzq7h0m7i04liyzrvnng1pi2-source/modules.nix:869:7:
          868|     optionalValue =
          869|       if isDefined then { value = mergedValue; }
             |       ^
          870|       else {};

       … while evaluating definitions from `/nix/store/714g6clapigzz83jh4gb8r4q9zzndk10-source/modules/transposition.nix':

       … while evaluating definitions from `/nix/store/smkvydkb12kwq35jxf1kh9g7h49z1cnq-source/flake-module.nix, via option perSystem':

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: evaluation aborted with the following error message: 'lib.customisation.callPackageWith: Function called without required argument "a" at /nix/store/p18pwjzr1bywzb84md8ira3z3c08ssdf-source/pkgs-by-name/b/package.nix:1, did you mean "R", "a4" or "ao"?'
```

With the repo set up like this:

```
└── pkgs-by-name
    ├── a
    │   └── package.nix
    └── b
        └── package.nix
```

```
a/package.nix
```
```nix
{ hello }: hello
```

```
b/package.nix
```
```nix
{ a }: a
```

### PR DESCRIPTION
**Pull request by xieve**
Closes #1.

- Refactored the flattening function to be a lot simpler
- Moved the `packagesFromDirectoryRecursive` call into the `makeScope` call, thereby allowing the definition of interdependent packages in a `pkgsDirectory`

There is one very particular edge case that my code handles worse than the old one: If a user had passed a non-nixpkgs `lib` where `callPackage` is broken in a way that doesn't always return derivations, unexpected things will happen: If the return value is not a set, it will be silently ignored and not included (because of the way I decided to filter the `mkScope` return value). The old implementation would have thrown an appropriate error. I could handle this more specifically, but given the incredibly narrow nature of this edge case, I would prefer to keep the code a little less ugly.

### CODE REVIEW

**Comment by drupol** [comment]
Hello,

Thanks for working on this! I’ll need a bit of time to fully understand your changes, as I’m not yet completely comfortable with everything.  

I appreciate the idea of calling `lib.filesystem.packagesFromDirectoryRecursive` only once, instead of twice as in the current implementation.  

However, one concern I have is that the `flatten` function and the call to `filterAttrs` relies on `config`, which is not passed as a parameter. While this works technically, I prefer explicit functions that only depend on their explicit inputs rather than referencing external variables that aren’t in scope.  

Would you mind documenting as much as possible the limitations and edge case introduced by this change in the README?  

Thanks again!

**Comment by xieve** [comment]
I'm already working on the README :)

> However, one concern I have is that the flatten function and the call to filterAttrs relies on config, which is not passed as a parameter. While this works technically, I prefer explicit functions that only depend on their explicit inputs rather than referencing external variables that aren’t in scope.

I see what you mean, and I did actually contemplate whether I should do it this way, but since all of this is scoped to the `perSystem` function specifically and `flattenPkgs` is not reusable anyway (we're also calling `isDerivation` as the stop condition), I thought that it would add unnecessary complexity. The only reason `flattenPkgs` has a name at all, as I see it, is because it's recursive. I generally try to abstract things only if I would repeat myself otherwise.
If you still prefer it, I can make that change though :)

**Comment by drupol** [comment]
> I'm already working on the README :)

Cool!

> I see what you mean, and I did actually contemplate whether I should do it this way, but since all of this is scoped to the `perSystem` function specifically and `flattenPkgs` is not reusable anyway (we're also calling `isDerivation` as the stop condition), I thought that it would add unnecessary complexity.

Just out of curiosity, could you just do it (in an isolated commit) so we can compare and decide later to include or remove the commit?

> The only reason `flattenPkgs` has a name at all, as I see it, is because it's recursive. I generally try to abstract things only if I would repeat myself otherwise. If you still prefer it, I can make that change though :)

This is a valid reason, I fully agree.

**Comment by xieve** [comment]
> Just out of curiosity, could you just do it (in an isolated commit) so we can compare and decide later to include or remove the commit?

Here you go: https://github.com/xieve/pkgs-by-name-for-flake-parts/commit/9246959ecd192e0191c866287726c9468139396d

**Comment by drupol** [comment]
> > Just out of curiosity, could you just do it (in an isolated commit) so we can compare and decide later to include or remove the commit?
> 
> Here you go: [xieve@9246959](https://github.com/xieve/pkgs-by-name-for-flake-parts/commit/9246959ecd192e0191c866287726c9468139396d)

Thanks for doing it, much appreciated. I prefer this method indeed, this is clearer for me to see what exactly the function need to work properly. Are you OK to include that commit in here?

I see you also fixed the edge case, this is great!

**Comment by drupol** [comment]
Is this PR fully ready now? Now that you have fixed the edge case, do you still need to modify the README or... ?

**Comment by xieve** [comment]
> Is this PR fully ready now? Now that you have fixed the edge case, do you still need to modify the README or... ?

I'm still working on the README, yes, and I should also definitely give this another test run. I'll let you know when it's ready :)

**Comment by xieve** [comment]
Oh and one more thing: I'm very unsure about the performance of this. `makeScope` is pretty likely to consume performance imo, since it involves fixed point calculations, but I have neither a good setup to profile Nix nor have I a repository that would be suitable to test this on. I do have a few ideas how it could possibly be improved, but I don't think it would make sense to implement them "just in case" without being able to verify whether they yield any improvement, especially since Nix is a very abstract language.

**Comment by drupol** [comment]
> Oh and one more thing: I'm very unsure about the performance of this. `makeScope` is pretty likely to consume performance imo, since it involves fix point calculations, but I have neither a good setup to profile Nix nor have I a repository that would be suitable to test this on. I do have a few ideas how it could possibly be improved, but I don't think it would make sense to implement them "just in case" without being able to verify whether they yield any improvement, especially since Nix is a very abstract language.

Maybe we can ask @roberth and see if he has a point of view on this?

**Comment by xieve** [comment]
I've hit a non-trivial roadblock: This does not seem to work well with nested folders. Specifically, I can't seem to get callPackage to pass an adjacent package that is located in a subfolder. I thought that this should be possibly like this:
```nix
{ directoryName }: mkDerivation { buildInputs = [ directoryName.packageName ]; ... }
```
but that does not seem to be the case. I'll investigate.

**Comment by xieve** [comment]
I was testing on upstream (I copied the example flake.nix) xD
It's all working now, you can merge :)

File: README.md [hunk 1]
```diff
@@ -80,52 +91,91 @@ Given this `flake.nix` file:
...
-        └── c
-            └── package.nix
+.
+├── flake.lock
+├── flake.nix
+└── packages
+    ├── pkg1
+    │   └── package.nix
+    └── subdirectory
+        └── pkg2.nix
```
Comment (drupol) [inline README.md:111]: Shouldn't this file be named package.nix?

File: README.md [hunk 2]
```diff
@@ -3,54 +3,60 @@
...
 The structure of the directory containing the packages
 is the same structure as the `pkgs/by-name` directory from `nixpkgs`.
 
-The first step is to import the input.
+The first step is to add this module as an input to your flake.
 
 ```nix
   inputs = {
-      pkgs-by-name-for-flake-parts.url = "github:drupol/pkgs-by-name-for-flake-parts";
+    pkgs-by-name-for-flake-parts.url = "github:drupol/pkgs-by-name-for-flake-parts";
```
Comment (drupol) [inline README.md:20]: Let's keep the indentation to 2 perhaps?

File: README.md [hunk 3]
```diff
@@ -3,54 +3,60 @@
...
   };
 ```
 
 Then, import the module:
 
 ```nix
   imports = [
-      inputs.pkgs-by-name-for-flake-parts.flakeModule
+    inputs.pkgs-by-name-for-flake-parts.flakeModule
```
Comment (drupol) [inline README.md:28]: Here as well

File: README.md [hunk 4]
```diff
@@ -3,54 +3,60 @@
...
   ];
 ```
 
-Then configure the module:
+and configure it:
 
 ```nix
   perSystem = {
-      pkgsDirectory = ./nix/pkgs;
+    pkgsDirectory = ./nix/pkgs;
```
Comment (drupol) [inline README.md:36]: Here too

**Comment by drupol** [comment]
Thank you for this contribution <3, it's still a bit complicated for me to understand everything there, but I tested it, and it's working better than before, so I merged it. I fixed the small nits myself.

File: README.md (hunk 1 above)
Comment (xieve) [reply README.md:111]: No, this is intentional. I explained it a few lines below :)

> Note how this package does not have its own directory. You can have either
`<package name>/package.nix`, or `<package name>.nix`. The former is useful if
you want to split the package into multiple nix files.

File: README.md (hunk 2 above)
Comment (xieve) [reply README.md:20]: I removed two spaces so there would be two left

Comment (drupol) [reply README.md:20]: Indeed, you were right actually.

**Comment by xieve** [comment]
Cleanup commit looks good :)
Let me know if you have any specific question, maybe I can leave some more comments so other people can understand it. It took a while to figure this out for me as well lol

**Comment by drupol** [comment]
I don't know if this is related, but I'm trying to update my system (from [`drupol/nixos-x260`](https://github.com/drupol/nixos-x260)) and I got this issue:

```
❯ nix flake update
❯ nh os switch .
> Building NixOS configuration
error:
       … while evaluating the attribute 'optionalValue.value'
         at /nix/store/9jn1cxgvdiv2jd7b8g794qiv16xxv7bq-source/modules.nix:893:5:
          892|
          893|     optionalValue =
             |     ^
          894|       if isDefined then { value = mergedValue; }

       … while evaluating a branch condition
         at /nix/store/9jn1cxgvdiv2jd7b8g794qiv16xxv7bq-source/modules.nix:894:7:
          893|     optionalValue =
          894|       if isDefined then { value = mergedValue; }
             |       ^
          895|       else {};

       … while evaluating definitions from `/nix/store/iskp8axrkjgrhc3mr6n6b2acdam3vkhv-source/modules/transposition.nix':

       … while evaluating the option `perSystem.x86_64-linux.packages':

       … while evaluating definitions from `/nix/store/8gn4xb7mzs5sbigpnz3qrh3awbz824g2-source/flake-module.nix, via option perSystem':

       … while evaluating the module argument `pkgs' in ":anon-17:anon-1":

       … while evaluating definitions from `/nix/store/rq2q5qgrwvf72400i92sags0jk5sd9c0-source/imports/overlay.nix, via option perSystem':

       … while evaluating definitions from `/nix/store/iskp8axrkjgrhc3mr6n6b2acdam3vkhv-source/extras/easyOverlay.nix':

       … while evaluating the option `perSystem.x86_64-linux.overlayAttrs':

       … while evaluating definitions from `/nix/store/rq2q5qgrwvf72400i92sags0jk5sd9c0-source/imports/pkgs.nix, via option perSystem':

       (641 duplicate frames omitted)

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: stack overflow; max-call-depth exceeded
       at /nix/store/j33wzkzndh41cyyy7i18bqm1srlv84cq-source/lib/systems/parse.nix:487:17:
          486|     "4" =    { cpu = elemAt l 0; vendor = elemAt l 1; kernel = elemAt l 2; abi = elemAt l 3; };
          487|   }.${toString (length l)}
             |                 ^
          488|     or (throw "system string '${lib.concatStringsSep "-" l}' has invalid number of hyphen-separated components");
┏━ 1 Errors: 
 ⋮ 
┃           895|       else {};
┃ 
┃        … while evaluating definitions from `/nix/store/iskp8axrkjgrhc3mr6n6b2acdam3vkhv-source/modules/transposition.nix':
┃ 
┃        … while evaluating the option `perSystem.x86_64-linux.packages':
┃ 
┃        … while evaluating definitions from `/nix/store/8gn4xb7mzs5sbigpnz3qrh3awbz824g2-source/flake-module.nix, via option perSystem':
┃ 
┃        … while evaluating the module argument `pkgs' in ":anon-17:anon-1":
┃ 
┃        … while evaluating definitions from `/nix/store/rq2q5qgrwvf72400i92sags0jk5sd9c0-source/imports/overlay.nix, via option perSystem':
┃ 
┃        … while evaluating definitions from `/nix/store/iskp8axrkjgrhc3mr6n6b2acdam3vkhv-source/extras/easyOverlay.nix':
┃ 
┃        … while evaluating the option `perSystem.x86_64-linux.overlayAttrs':
┃ 
┃        … while evaluating definitions from `/nix/store/rq2q5qgrwvf72400i92sags0jk5sd9c0-source/imports/pkgs.nix, via option perSystem':
┃ 
┃        (641 duplicate frames omitted)
┃ 
┃        (stack trace truncated; use '--show-trace' to show the full, detailed trace)
┃ 
┃        error: stack overflow; max-call-depth exceeded
┃        at /nix/store/j33wzkzndh41cyyy7i18bqm1srlv84cq-source/lib/systems/parse.nix:487:17:
┃           486|     "4" =    { cpu = elemAt l 0; vendor = elemAt l 1; kernel = elemAt l 2; abi = elemAt l 3; };
┃           487|   }.${toString (length l)}
┃              |                 ^
┃           488|     or (throw "system string '${lib.concatStringsSep "-" l}' has invalid number of hyphen-separated components");
┣━━━                                                            
┗━ ∑ ⚠ Exited with 1 errors reported by nix at 11:52:59 after 1s
Error: 
   0: Command exited with status Exited(1)

Location:
   src/commands.rs:151
~/C/d/nixos-x260 >
```

Investigating...

**Comment by xieve** [comment]
Perhaps this update is conflicting with your workaround for #1 via `easyOverlay`?

**Comment by drupol** [comment]
When I remove the easyOverly module and the workarounds, I have this issue now:

```
❯ nh os switch .
> Building NixOS configuration
error:
       … while calling the 'head' builtin
         at /nix/store/j33wzkzndh41cyyy7i18bqm1srlv84cq-source/lib/attrsets.nix:1574:11:
         1573|         || pred here (elemAt values 1) (head values) then
         1574|           head values
             |           ^
         1575|         else

       … while evaluating the attribute 'value'
         at /nix/store/j33wzkzndh41cyyy7i18bqm1srlv84cq-source/lib/modules.nix:846:9:
          845|     in warnDeprecation opt //
          846|       { value = addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          847|         inherit (res.defsFinal') highestPrio;

       … while evaluating the option `system.build.toplevel':

       … while evaluating definitions from `/nix/store/j33wzkzndh41cyyy7i18bqm1srlv84cq-source/nixos/modules/system/activation/top-level.nix':

       … while evaluating the option `warnings':

       … while evaluating definitions from `/nix/store/j33wzkzndh41cyyy7i18bqm1srlv84cq-source/nixos/modules/system/boot/systemd.nix':

       … while evaluating the option `systemd.services.home-manager-pol.serviceConfig':

       … while evaluating definitions from `/nix/store/psywlz2kzak45iy3x9nrss6vp0xz55mk-source/hosts/common/system/imports.nix':

       … while evaluating the option `home-manager.users.pol.home.file.".manpath".source':

       … while evaluating definitions from `/nix/store/n16ia1vwgl5n4kqf5qdlx2z0mr8jjd5x-source/modules/files.nix':

       … while evaluating the option `home-manager.users.pol.home.file.".manpath".text':

       … while evaluating definitions from `/nix/store/n16ia1vwgl5n4kqf5qdlx2z0mr8jjd5x-source/modules/programs/man.nix':

       … while evaluating the option `home-manager.users.pol.home.packages':

       … while evaluating definitions from `/nix/store/psywlz2kzak45iy3x9nrss6vp0xz55mk-source/modules/home/messaging/default.nix':

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: undefined variable 'chromium-discord'
       at /nix/store/psywlz2kzak45iy3x9nrss6vp0xz55mk-source/modules/home/messaging/default.nix:17:7:
           16|     home.packages = with pkgs; [
           17|       chromium-discord
             |       ^
           18|       chromium-element
┏━ 1 Errors: 
 ⋮ 
┃        … while evaluating the option `warnings':
┃ 
┃        … while evaluating definitions from `/nix/store/j33wzkzndh41cyyy7i18bqm1srlv84cq-source/nixos/modules/system/boot/systemd.nix':
┃ 
┃        … while evaluating the option `systemd.services.home-manager-pol.serviceConfig':
┃ 
┃        … while evaluating definitions from `/nix/store/psywlz2kzak45iy3x9nrss6vp0xz55mk-source/hosts/common/system/imports.nix':
┃ 
┃        … while evaluating the option `home-manager.users.pol.home.file.".manpath".source':
┃ 
┃        … while evaluating definitions from `/nix/store/n16ia1vwgl5n4kqf5qdlx2z0mr8jjd5x-source/modules/files.nix':
┃ 
┃        … while evaluating the option `home-manager.users.pol.home.file.".manpath".text':
┃ 
┃        … while evaluating definitions from `/nix/store/n16ia1vwgl5n4kqf5qdlx2z0mr8jjd5x-source/modules/programs/man.nix':
┃ 
┃        … while evaluating the option `home-manager.users.pol.home.packages':
┃ 
┃        … while evaluating definitions from `/nix/store/psywlz2kzak45iy3x9nrss6vp0xz55mk-source/modules/home/messaging/default.nix':
┃ 
┃        (stack trace truncated; use '--show-trace' to show the full, detailed trace)
┃ 
┃        error: undefined variable 'chromium-discord'
┃        at /nix/store/psywlz2kzak45iy3x9nrss6vp0xz55mk-source/modules/home/messaging/default.nix:17:7:
┃            16|     home.packages = with pkgs; [
┃            17|       chromium-discord
┃              |       ^
┃            18|       chromium-element
┣━━━                                                            
┗━ ∑ ⚠ Exited with 1 errors reported by nix at 12:03:41 after 4s
Error: 
   0: Command exited with status Exited(1)

Location:
   src/commands.rs:151
~/C/d/nixos-x260 > (~ (no description set) tnnv 0da0) ✘    
```

I'm still investigating on how to get it working again.

**Comment by xieve** [comment]
Ah yeah, since you now don't have an overlay, you'll have to pass your packages to your modules in another way. 
For example:
```nix
# flake.nix
{
  inputs = { /* ... */ };
  outputs = inputs@{ ... }:
    inputs.flake-parts.lib.mkFlake { inherit inputs; } {
      flake.nixosModules.myModule = import ./my-module.nix { inherit inputs; };
    }
}
```
```nix
# my-module.nix
{ inputs }:
{
  config.environment.systemPackages = [ inputs.self.packages.chromium-discord ];
}
```

**Comment by xieve** [comment]
Alternatively, there's probably also a way to re-add the overlay. I'm not sure how `easyOverlay` works but as long as your overlay is not in scope for the package evaluation but only for your nixos modules that should work

**Comment by drupol** [comment]
I made it work using your proposal: https://github.com/drupol/nixos-x260/commit/331651ff85866bf250b0e8b8e33562e905da0350

I don't like it so much, I wish I could find a way to get it working through the `pkgs` variable, it was much easier.

**Comment by drupol** [comment]
If you have any idea on how to make an overlay, please let me know. I'm looking into this thing since more than an hour without luck so far.

**Comment by drupol** [comment]
There's an issue when doing:

```nix
{ ... }:
{

  perSystem =
    { config, ... }:
    {
      pkgsDirectory = ./pkgs;
      packages = config.packages
    };
}
```

<details>

<summary>error log</summary>

```
❯ nix flake show
git+file:///home/pol/Code/tmp/flow
├───devShells
│   ├───aarch64-darwin
│   │   └───default omitted (use '--all-systems' to show)
│   ├───aarch64-linux
│   │   └───default omitted (use '--all-systems' to show)
│   ├───x86_64-darwin
│   │   └───default omitted (use '--all-systems' to show)
│   └───x86_64-linux
│       └───default: development environment 'nix-shell'
├───legacyPackages
│   ├───aarch64-darwin omitted (use '--legacy' to show)
│   ├───aarch64-linux omitted (use '--legacy' to show)
│   ├───x86_64-darwin omitted (use '--legacy' to show)
│   └───x86_64-linux omitted (use '--legacy' to show)
└───packages
    ├───aarch64-darwin
error:
       … while evaluating the attribute 'optionalValue.value'
         at /nix/store/9jn1cxgvdiv2jd7b8g794qiv16xxv7bq-source/modules.nix:893:5:
          892|
          893|     optionalValue =
             |     ^
          894|       if isDefined then { value = mergedValue; }

       … while evaluating a branch condition
         at /nix/store/9jn1cxgvdiv2jd7b8g794qiv16xxv7bq-source/modules.nix:894:7:
          893|     optionalValue =
          894|       if isDefined then { value = mergedValue; }
             |       ^
          895|       else {};

       … while evaluating definitions from `/nix/store/iskp8axrkjgrhc3mr6n6b2acdam3vkhv-source/modules/transposition.nix':

       … while evaluating the option `perSystem.aarch64-darwin.packages':

       … while evaluating definitions from `/nix/store/gvqcyr3ablj6qvascayiqrcbbp6r6079-source/.nix/pkgs.nix, via option perSystem':

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: infinite recursion encountered
       at /nix/store/9jn1cxgvdiv2jd7b8g794qiv16xxv7bq-source/modules.nix:846:9:
          845|     in warnDeprecation opt //
          846|       { value = addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          847|         inherit (res.defsFinal') highestPrio;
```

</details>


However, if I do:

```nix
{ ... }:
{

  perSystem =
    { config, ... }:
    {
      pkgsDirectory = ./pkgs;
      packages.foobar = config.packages.foobar
    };
}
```

It works well.

I'm still investigating on how we can expose local packages through an overlay or any other mechanism. If you have a clue, let me know please.

**Comment by drupol** [comment]
Ignore my last comment, my brain made a fart on this.

**Comment by xieve** [comment]
Did you figure it out? IMO it would make sense if this module provided an overlay you could include or maybe an option to automatically include an overlay in every NixOS config, but I haven't gotten around to looking into it yet.

**Comment by drupol** [comment]
The current module already sets `packages`, which means all local packages are exposed and visible when running `nix flake show`. However, what I'm looking for is a way to expose and overlay them, but I haven't found a clean solution yet...

**Comment by xieve** [comment]
Figured it out :D
This should be a drop-in replacement for your dotfiles with https://github.com/drupol/nixos-x260/commit/331651ff85866bf250b0e8b8e33562e905da0350 reverted:
```nix
mkFlake { inherit inputs; } { inputs, withSystem, ... }: {
  flake.overlays.default = final: prev:
    withSystem prev.stdenv.hostPlatform.system (
      { config, ... }: config.packages
    );
}
```
Although I prefer
```nix
mkFlake { inherit inputs; } { inputs, withSystem, ... }: {
  flake.overlays.default = final: prev:
    withSystem prev.stdenv.hostPlatform.system (
      { config, ... }: {
        xieve = config.packages;
      }
    );
}
```
to prevent any accidental collisions and for clarity. See also the [flake-parts docs on overlays](https://flake.parts/overlays.html#defining-an-overlay).

This would probably be a good addition to the README.

**Comment by drupol** [comment]
Hey!

Thanks for coming back to this :)

When I try this, I get:

```
error:
       … while evaluating the attribute 'optionalValue.value'
         at /nix/store/nb5dk7rsj1dzww7n52gfdbrxk380dkh2-source/modules.nix:974:5:
          973|
          974|     optionalValue =
             |     ^
          975|       if isDefined then { value = mergedValue; }

       … while evaluating a branch condition
         at /nix/store/nb5dk7rsj1dzww7n52gfdbrxk380dkh2-source/modules.nix:975:7:
          974|     optionalValue =
          975|       if isDefined then { value = mergedValue; }
             |       ^
          976|       else {};

       … while evaluating definitions from `/nix/store/054s12i2a5vg9wxp308ns2qxb38zbbz4-source/modules/transposition.nix':

       … while evaluating definitions from `/nix/store/fvn44m1ml2k7946r5z53j6ya28pgwgi9-source/imports/devshells.nix, via option perSystem':

       … while evaluating the module argument `pkgs' in ":anon-14:anon-1":

       … while evaluating definitions from `/nix/store/fvn44m1ml2k7946r5z53j6ya28pgwgi9-source/imports/overlay.nix, via option perSystem':

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: The option `flake.overlays.default' is defined multiple times while it's expected to be unique.

       Definition values:
       - In `/nix/store/054s12i2a5vg9wxp308ns2qxb38zbbz4-source/extras/easyOverlay.nix': <function>
       - In `/nix/store/fvn44m1ml2k7946r5z53j6ya28pgwgi9-source/imports/overlay.nix': <function>
       Use `lib.mkForce value` or `lib.mkDefault value` to change the priority on any of these definitions.
~
```

**Comment by xieve** [comment]
> Hey!
> 
> Thanks for coming back to this :)
> 
> When I try this, I get:
> 
>\<snip>

You still need to remove `easyOverlay` from your flake imports I think

**Comment by drupol** [comment]
Nope:

```
error:
       … while evaluating the attribute 'optionalValue.value'
         at /nix/store/nb5dk7rsj1dzww7n52gfdbrxk380dkh2-source/modules.nix:974:5:
          973|
          974|     optionalValue =
             |     ^
          975|       if isDefined then { value = mergedValue; }

       … while evaluating a branch condition
         at /nix/store/nb5dk7rsj1dzww7n52gfdbrxk380dkh2-source/modules.nix:975:7:
          974|     optionalValue =
          975|       if isDefined then { value = mergedValue; }
             |       ^
          976|       else {};

       … while evaluating definitions from `/nix/store/054s12i2a5vg9wxp308ns2qxb38zbbz4-source/modules/transposition.nix':

       … while evaluating definitions from `/nix/store/mp8rj671dx4x2wki4il6c4l56fh4qlmh-source/imports/devshells.nix, via option perSystem':

       … while evaluating the module argument `pkgs' in ":anon-13:anon-1":

       … while evaluating definitions from `/nix/store/mp8rj671dx4x2wki4il6c4l56fh4qlmh-source/imports/overlay.nix, via option perSystem':

       … while evaluating definitions from `/nix/store/mp8rj671dx4x2wki4il6c4l56fh4qlmh-source/imports/overlay.nix':

       … while evaluating the option `perSystem.x86_64-linux.packages':

       … while evaluating definitions from `/nix/store/7zn3hx70y02d5zbmga290lv6kmg9qm4a-source/flake-module.nix, via option perSystem':

       … while evaluating the module argument `pkgs' in ":anon-16:anon-1":

       … while evaluating definitions from `/nix/store/mp8rj671dx4x2wki4il6c4l56fh4qlmh-source/imports/overlay.nix, via option perSystem':

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: infinite recursion encountered
       at /nix/store/nb5dk7rsj1dzww7n52gfdbrxk380dkh2-source/types.nix:666:71:
          665|       # Push down position info.
          666|       pushPositions = map (def: mapAttrs (n: v: { inherit (def) file; value = v; }) def.value);
             |                                                                       ^
          667|       binOp = lhs: rhs:
~/C/d/nixos-x260 > (~master (no description set) ttlx 88a0 (empty)) ✘                                                                                  
```

**Comment by drupol** [comment]
Got it working!

Done @ https://github.com/drupol/nixos-x260/commit/23103c0b66989faf88ff34e9cd078aa93c2466d5

Thanks !

**Comment by xieve** [comment]
Ah, it's possible that easy-overlay never was the problem at all, I didn't try it myself. What I neglected to mention is that I consume my overlay at NixOS configuration level, not at flake level: [common.nix](https://github.com/xieve/dotfiles/blob/a970e9f02a3a78c319178c27fa65d542d997142f/nixos/common.nix#L22-L25)

The issue is that pkgs-by-name-for-flake-parts is called with `_module.args` (because it is based on perSystem), ie. whatever you specify in there has to be evaluated for the packages to be evaluated, so if you put your own packages in there you get an infinite recursion error. An alternative approach could be to create a subflake for the packages which provides an overlay, but I don't need to consume my overlay at flake level, so I just consume the overlay inside a module that is loaded by all configurations.

**Comment by xieve** [comment]
Ah, only saw your comment now, nice!!

**Comment by drupol** [comment]
I just updated the README, feel free to let me know what you think, and feel free to submit a PR to fix things.

Thanks again for this nice improvements!

**Comment by drupol** [comment]
Hello,

Sorry to dig up old topics, but I just saw this: https://infosec.exchange/@katexochen/114230458727016712

Related PR: https://github.com/NixOS/nixpkgs/pull/392800

Do you think we will have to update something in the changes you implemented?

**Comment by drupol** [comment]
@xieve Do you have some feedback to share concerning that ^^
